x/gamm/keeper/pool_service_test.go

```diff
@@ -222,6 +242,187 @@ func (suite *KeeperTestSuite) TestCreateBalancerPool() {
+			emptySender: false,
+			expectPass:  true,
+		},
+		{
+			name:     "initialize a CL pool which cause panic",
+			isCLPool: true,
+		},
+	}
+
+	for _, test := range tests {
+		test := test
+		suite.Run(test.name, func() {
+			suite.SetupTest()
+
+			gammKeeper := suite.App.GAMMKeeper
+			swaprouterKeeper := suite.App.SwapRouterKeeper
+			bankKeeper := suite.App.BankKeeper
+			poolIncentivesKeeper := suite.App.PoolIncentivesKeeper
+			communityPoolKeeper := suite.App.DistrKeeper
+
+			// fund sender test account
+			var sender sdk.AccAddress
+			if test.msg == nil {
+				sender = testAccount
+			} else {
+				sender = test.msg.PoolCreator()
+
+			}
+			if !test.emptySender {
+				suite.FundAcc(sender, defaultAcctFunds)
+			}
+
+			// note starting balances for community fee pool and pool creator account
+			senderBalBeforeNewPool := bankKeeper.GetAllBalances(suite.Ctx, sender)
+
+			if test.isCLPool {
+				ctrl := gomock.NewController(suite.T())
+				mockPool := mocks.NewMockConcentratedPoolExtension(ctrl)
+				mockPool.EXPECT().GetTotalShares().Return(sdk.NewInt(100)).AnyTimes()
+				mockPool.EXPECT().GetId().Return(uint64(1)).AnyTimes()
+				suite.Require().Panics(func() { gammKeeper.InitializePool(suite.Ctx, mockPool, sender) })
+			} else {
+				// createPool process
+				err := test.msg.Validate(suite.Ctx)
+				suite.Require().NoError(err, "test: %v", test.name)
+
+				// note starting balance of pool
+				poolBalanceBefore := sdk.Coins{}
+				for _, asset := range test.msg.InitialLiquidity() {
+					poolBalanceBefore = poolBalanceBefore.Add(sdk.NewCoin(asset.Denom, gammKeeper.GetDenomLiquidity(suite.Ctx, asset.Denom)))
+				}
+
+				// create pool process which happens previous the initializePool phase
+				initialPoolLiquidity := test.msg.InitialLiquidity()
+
+				numAssets := initialPoolLiquidity.Len()
+				suite.Require().Equal(numAssets < swaproutertypes.MinPoolAssets, false)
+				suite.Require().Equal(numAssets > swaproutertypes.MaxPoolAssets, false)
+
+				fee := swaprouterKeeper.GetParams(suite.Ctx).PoolCreationFee
+				err = communityPoolKeeper.FundCommunityPool(suite.Ctx, fee, sender)
+				suite.Require().NoError(err, "test: %v", test.name)
+
+				poolId := swaprouterKeeper.GetNextPoolId(suite.Ctx)
+				swaprouterKeeper.SetNextPoolId(suite.Ctx, poolId+1)
+
+				pool, err := test.msg.CreatePool(suite.Ctx, poolId)
+				suite.Require().NoError(err, "test: %v", test.name)
+
+				swaprouterKeeper.SetPoolRoute(suite.Ctx, poolId, test.msg.GetPoolType())
+				err = osmoutils.CreateModuleAccount(suite.Ctx, suite.App.AccountKeeper, pool.GetAddress())
+				suite.Require().NoError(err, "test: %v", test.name)
+
+				poolAddr := pool.GetAddress()
+				err = bankKeeper.SendCoins(suite.Ctx, sender, poolAddr, initialPoolLiquidity)
+				suite.Require().NoError(err, "test: %v", test.name)
+
+				swapModule, err := swaprouterKeeper.GetPoolModule(suite.Ctx, poolId)
+				suite.Require().NoError(err, "test: %v", test.name)
```

Comment:
I don't see why all this setup logic is needed.

We should be able to simply do the following:
- define `swaprouter.PoolI` as a filed on the test case
- for each test case., initialize each pool via their respective constructor such as 
    * balancer: https://github.com/osmosis-labs/osmosis/blob/6436e5200ac38a3f39160f12b5c9fda5e57df7bd/x/gamm/pool-models/balancer/pool.go#L43
    * stableswap: https://github.com/osmosis-labs/osmosis/blob/6436e5200ac38a3f39160f12b5c9fda5e57df7bd/x/gamm/pool-models/stableswap/pool.go#L25
    * CL: https://github.com/osmosis-labs/osmosis/blob/6436e5200ac38a3f39160f12b5c9fda5e57df7bd/x/concentrated-liquidity/model/pool.go#L20

Then, all this swap router-related boilerplate becomes obsolete and we can simply provide a pool struct to `InitializePool` method below.

Please see this similar test for inspiration: https://github.com/osmosis-labs/osmosis/blob/6436e5200ac38a3f39160f12b5c9fda5e57df7bd/x/concentrated-liquidity/pool_test.go#L78

Reply:
I have remove all the setup logic and use the constructors to initialize pool. Hope this time it looks good to you. Can you check it again ? @p0mvn